enhancements/separating-provider-specific-code-in-the-installer.md

```diff
@@ -0,0 +1,147 @@
+---
+title: separating-provider-specific-code-in-the-installer
+authors:
+  - "@janoszen"
+reviewers:
+  - "@Gal-Zaidman"
+approvers:
+  - TBD
+creation-date: 2020-11-27
+last-updated: 2020-11-27
+status: provisional
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Separating provider-specific code in the installer
+
+## Release Signoff Checklist
+
+TBD
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+This change introduces an internal API in the installer code to move provider-specific code into a separate folder and avoid having large switch-case structures in the installer code.
+
+## Motivation
+
+Several parts of the installer code have large, [unwieldy switch-case structures](https://github.com/openshift/installer/blob/254680f6268a7b8e57091d788256c67b81755b5d/pkg/asset/cluster/tfvars.go#L166) that contain provider-specific codes. This makes it hard to keep track of the code that serves a provider, and it makes it impossible to write component-level tests. Creating a separated API reduces friction between teams and makes changes easier to merge because the risk of accidental side effects between providers is reduced. 
```

Comment:
> makes it impossible to write component-level tests

What do you mean by "component-level tests"? Many platforms have platform-specific unit tests. For example, here are [the GCP tests for validating install configs](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/gcp/validation_test.go).

Reply:
We (oVirt) have several parts of our code where half of the functionality is in a switch-case, the other half is in a switch-case. It would be helpful to have tests that run our code only without needing to test the whole installer. That makes test runs quicker and less flaky.

Reply:
You can't really call them unit tests because they still require external API's, which is a separate issue to be addressed.

Reply:
If you look at the GCP example I linked above you can see how we mocked out the GCP APIs.